### PR TITLE
lesson - 4

### PR DESCRIPTION
**Pull request by DariaRomanowska**

### CODE REVIEW

File: src/index.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+	<meta name="keywords" content="HTML, CSS, JavaScript" />
+	<meta name="viewport" content="width=device-width, initial-scale=1.0" />
+</head>
+<body>
+<h1>Hello world!</h1>
+
+<p style="cursor: pointer; padding: 5px; font-size: 24px; background-color: #000000" id="text1">Text 1</p>
+<p style="cursor: pointer; padding: 5px; font-size: 24px; background-color: #000000" id="text2">Text 2</p>
+<p style="cursor: pointer; padding: 5px; font-size: 24px; background-color: #000000" id="text3">Text 3</p>
+<script type="module" src="scripts/main.js"></script>
```
Comment (5tronciy) [inline src/index.html]: Do you need module type here? Unlike classic scripts, module scripts require the use of the CORS protocol for cross-origin fetching.
"text/javascript" should work well.

File: src/scripts/main.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+console.log('----------------');
+
+const text1 = document.getElementById('text1');
+const text2 = document.getElementById('text2');
+const text3 = document.getElementById('text3');
+
+const colors = ['magenta', 'cyan', 'firebrick', 'springgreen', 'skyblue'];
+let i = 0;
+
+function myFunction() {
```
Comment (5tronciy) [inline src/scripts/main.js]: The name should say to other people what the function is doing.

**Review by 5tronciy — CHANGES_REQUESTED** [review]
Clear the branch from files and project from dependencies copied from another project. Leave what you need for the task only.
You may use prettier, linter and other dev tools.
Add required tests for your functions.
And the requirement of independent color changing for different blocks is not met.

File: src/index.html (hunk 1 above)
Comment (DariaRomanowska) [reply src/index.html]: I wrote it as in the video tutorial when it was necessary to connect export / import

Comment (5tronciy) [reply src/index.html]: Where did you use export/import in your project?

File: src/scripts/module.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  return typeof price === 'number';
+}
+test('When adding new valid product, get successful confirmation', () => {
+  const name = randProductName();
+  const price = randNumber();
+  // Generated random input: { 'Handmade Granite Cheese', 704809 }
+  expect(addProduct(name, price)).toBe(true);
+  // Test failed, the random input triggered some path we never planned for.
+  // We discovered a bug early!
+});
```
Comment (5tronciy) [inline src/scripts/module.js]: Where did you get this file? Is it required at the task?
And where is your tests?

File: package.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+  "dependencies": {
+    "save-dev": "^0.0.1-security"
```
Comment (5tronciy) [inline package.json:5]: What is it for here?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  "devDependencies": {
+    "@ngneat/falso": "^6.1.0",
```
Comment (5tronciy) [inline package.json:10]: Do you use it for the task?

**Review by 5tronciy — CHANGES_REQUESTED** [review]
The tests are still absent.
You don't use tools from first lesson.
Look at the file index.html - the code style leaves much to be desired.
Also, I don't find the folder coverage to be useful. I think you may remove it from git.

File: package.json (hunk 5 above)
Comment (DariaRomanowska) [reply package.json:10]: I think it must be, cause I need to add a unit tests

File: package.json (hunk 4 above)
Comment (DariaRomanowska) [reply package.json:5]: I don`t know what is it and how is appeared here. May I delete it?

**Review by 5tronciy — CHANGES_REQUESTED** [review]
Nothing is changed except the formatting since previous review.
The tests should be in the separate file with name ends on `test.js`, for example `main.test.js`. Create 1 test for function sum(), and 1 test for function getColor().
Oh, I missed your comments on review. So, package jest in dependencies is enough for tests.

**Review by 5tronciy — CHANGES_REQUESTED** [review]
GitHub doesn't allow me to leave comments in files coz there are too many files in one PR. Remove the folder `node_modules` and keep your repo clean, please.
We don't load tests with index.html. The only environment for applying tests is Jest.

File: src/scripts/test/getColor.test.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+const colors = require('../main.js')
```
Comment (5tronciy) [inline src/scripts/test/getColor.test.js]:
```suggestion
const colors = ["magenta", "cyan", "firebrick", "springgreen", "skyblue"];
```

Comment (5tronciy) [reply src/scripts/test/getColor.test.js]: The main.js applies after importing with all its plain code.

File: src/scripts/test/getColor.test.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+      }
+    }
+  } else {
+    return colors[0];
+  }
+};
+
+
+test("get next color from array", function () {
+  expect(getColor(0)).toBe(colors[0]);
```
Comment (5tronciy) [inline src/scripts/test/getColor.test.js]:
```suggestion
  expect(getColor("cyan")).toBe("firebrick");
```

Comment (5tronciy) [reply src/scripts/test/getColor.test.js]: Tests should be clear to understand.

Comment (5tronciy) [reply src/scripts/test/getColor.test.js]: Also the value 0 is unreachable from the code using the function getColor() in your app.

**Review by 5tronciy — APPROVED** [review]
Ok, pretty well.
Unit tests are easier than you would expect. And they can be very useful if you use them consciously.

**Comment by 5tronciy** [comment]
Look at `.gitignore` and fix the 2nd row.

**Pull request merged**